File: persistence/src/main/java/io/sparkled/persistence/sequence/SequencePersistenceService.java
package io.sparkled.persistence.sequence;

import io.sparkled.model.entity.*;
import io.sparkled.model.render.RenderedStagePropDataMap;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

public interface SequencePersistenceService {

    Sequence createSequence(Sequence sequence);

    List<Sequence> getAllSequences();

    Optional<Sequence> getSequenceById(int sequenceId);

    Optional<Stage> getStageBySequenceId(int sequenceId);

    Optional<SongAudio> getSongAudioBySequenceId(int sequenceId);

    List<SequenceChannel> getSequenceChannelsBySequenceId(int sequenceId);

    Optional<SequenceChannel> getSequenceChannelByUuid(int sequenceId, UUID uuid);

    RenderedStagePropDataMap getRenderedStagePropsBySequenceAndSong(Sequence sequence, Song song);

    Map<String, UUID> getSequenceStagePropUuidMapBySequenceId(int sequenceId);

    void saveSequence(Sequence sequence, List<SequenceChannel> sequenceChannels);

    void publishSequence(Sequence sequence, List<SequenceChannel> sequenceChannels, RenderedStagePropDataMap renderedStageProps);

    void deleteSequence(int sequenceId);
}

File: persistence/src/main/kotlin/io/sparkled/persistence/sequence/SequencePersistenceService.kt
package io.sparkled.persistence.sequence

import io.sparkled.model.entity.Sequence
import io.sparkled.model.entity.SequenceChannel
import io.sparkled.model.entity.Song
import io.sparkled.model.entity.SongAudio
import io.sparkled.model.entity.Stage
import io.sparkled.model.render.RenderedStagePropDataMap
import java.util.UUID

interface SequencePersistenceService {

    fun getAllSequences(): List<Sequence>

    fun getSequenceById(sequenceId: Int): Sequence?

    fun getStageBySequenceId(sequenceId: Int): Stage?

    fun getSongAudioBySequenceId(sequenceId: Int): SongAudio?

    fun getSequenceChannelsBySequenceId(sequenceId: Int): List<SequenceChannel>

    fun getSequenceChannelByUuid(sequenceId: Int, uuid: UUID): SequenceChannel?

    fun getRenderedStagePropsBySequenceAndSong(sequence: Sequence, song: Song): RenderedStagePropDataMap

    fun getSequenceStagePropUuidMapBySequenceId(sequenceId: Int): Map<String, UUID>

    fun saveSequence(sequence: Sequence, sequenceChannels: List<SequenceChannel>): Sequence

    fun publishSequence(
        sequence: Sequence,
        sequenceChannels: List<SequenceChannel>,
        renderedStageProps: RenderedStagePropDataMap
    )

    fun deleteSequence(sequenceId: Int)
}
